### ISSUE TITLE
failure on mingw64 + longdouble

### ISSUE THREAD

**Issue by kmx** [rurban/Cpanel-JSON-XS#57]
Hi,

when building Cpanel-JSON-XS-3.0210 on
- perl 5.22.1 **WITH longdouble** / 64bit MS Windows
- gcc-4.9.2 compiler

I got these failures:

```
t/00_load.t ................. ok
t/01_utf8.t ................. ok
t/02_error.t ................ ok
t/03_types.t ................ ok
t/04_dwiw_encode.t .......... ok
t/05_dwiw_decode.t .......... ok
t/06_pc_pretty.t ............ ok
t/07_pc_esc.t ............... ok
t/08_pc_base.t .............. ok
t/08_pc_base_nv.t ........... ok
t/09_pc_extra_number.t ...... ok
t/104_sortby.t .............. ok
t/105_esc_slash.t ........... ok
t/106_allow_barekey.t ....... ok
t/107_allow_singlequote.t ... ok
t/108_decode.t .............. ok
t/109_encode.t .............. ok
t/10_pc_keysort.t ........... ok
t/110_bignum.t .............. ok
t/112_upgrade.t ............. ok
t/113_overloaded_eq.t ....... ok
t/114_decode_prefix.t ....... ok
t/115_tie_ixhash.t .......... ok
t/116_incr_parse_fixed.t .... ok

#   Failed test 'inf -> null'
#   at t/117_numbers.t line 7.
#          got: '[inf]'
#     expected: '[null]'

#   Failed test 'nan -> null'
#   at t/117_numbers.t line 8.
#          got: '[nan]'
#     expected: '[null]'

#   Failed test '-inf -> null'
#   at t/117_numbers.t line 9.
#          got: '[-inf]'
#     expected: '[null]'

#   Failed test '-nan -> null'
#   at t/117_numbers.t line 10.
#          got: '[nan]'
#     expected: '[null]'

#   Failed test '-nan -> null'
#   at t/117_numbers.t line 11.
#          got: '[nan]'
#     expected: '[null]'

#   Failed test 'inf -> "inf"'
#   at t/117_numbers.t line 25.
#          got: '[inf]'
#     expected: '["1.#INF"]'

#   Failed test '-inf -> "-inf"'
#   at t/117_numbers.t line 26.
#          got: '[-inf]'
#     expected: '["-1.#INF"]'

#   Failed test 'nan -> "nan"'
#   at t/117_numbers.t line 29.
#                   '[nan]'
#     doesn't match '(?^:\[\"(-1.#IND|1.#QNAN)\"\])'

#   Failed test '-nan -> "-nan"'
#   at t/117_numbers.t line 30.
#                   '[nan]'
#     doesn't match '(?^:\[\"(-1.#IND|1.#QNAN)\"\])'

#   Failed test '-nan -> "-nan"'
#   at t/117_numbers.t line 31.
#                   '[nan]'
#     doesn't match '(?^:\[\"(-1.#IND|1.#QNAN)\"\])'

#   Failed test 'inf'
#   at t/117_numbers.t line 34.
#          got: '[inf]'
#     expected: '[1.#INF]'

#   Failed test '-inf'
#   at t/117_numbers.t line 35.
#          got: '[-inf]'
#     expected: '[-1.#INF]'

#   Failed test 'nan'
#   at t/117_numbers.t line 36.
#                   '[nan]'
#     doesn't match '(?^:\[(-1.#IND|1.#QNAN)\])'

#   Failed test '-nan'
#   at t/117_numbers.t line 37.
#                   '[nan]'
#     doesn't match '(?^:\[(-1.#IND|1.#QNAN)\])'

#   Failed test '-nan'
#   at t/117_numbers.t line 38.
#                   '[nan]'
#     doesn't match '(?^:\[(-1.#IND|1.#QNAN)\])'
# Looks like you failed 15 tests of 19.
t/117_numbers.t ............. 
Dubious, test returned 15 (wstat 3840, 0xf00)
Failed 15/19 subtests 
t/11_pc_expo.t .............. ok
t/12_blessed.t .............. ok
t/13_limit.t ................ ok
t/14_latin1.t ............... ok
t/15_prefix.t ............... ok
t/16_tied.t ................. ok
t/17_relaxed.t .............. ok
t/18_json_checker.t ......... ok
t/19_incr.t ................. ok
t/20_faihu.t ................ ok
t/20_unknown.t .............. skipped: JSON required for cross testing
t/21_evans.t ................ ok
t/22_comment_at_eof.t ....... ok
t/23_array_ctx.t ............ ok
t/24_freeze_recursion.t ..... ok
t/25_boolean.t .............. ok
t/52_object.t ............... ok
t/53_readonly.t ............. ok
t/54_stringify.t ............ skipped: JSON required for cross testing
t/55_modifiable.t ........... ok
t/96_interop.t .............. skipped: JSON::XS and JSON required for testing interop
t/96_mojo.t ................. ok
t/97_unshare_hek.t .......... ok
t/98_56only.t ............... ok
t/99_binary.t ............... ok
t/z_kwalitee.t .............. skipped: This test is only run for the module author
t/z_leaktrace.t ............. skipped: require Test::LeakTrace
t/z_meta.t .................. skipped: This test is only run for the module author
t/z_perl_minimum_version.t .. skipped: Author tests not required for installation
t/z_pod-coverage.t .......... skipped: This test is only run for the module author
t/z_pod-spell-mistakes.t .... skipped: This test is only run for the module author
t/z_pod-spelling.t .......... skipped: This test is only run for the module author
t/z_pod.t ................... ok
t/zero-mojibake.t ........... ok

Test Summary Report
-------------------
t/117_numbers.t           (Wstat: 3840 Tests: 19 Failed: 15)
  Failed tests:  1-15
  Non-zero exit status: 15
Files=59, Tests=1480, 12 wallclock secs ( 0.58 usr +  0.17 sys =  0.75 CPU)
Result: FAIL
```

Interesting is that same perl version but **WITHOUT longdouble** PASSes all `Cpanel::JSON::XS` tests.

If you have a MS Win machine you can get longdouble build from http://strawberryperl.com/beta.html to replicate this failure.

**Comment by rurban** [comment]
Interesting thanks. I'll boot up my win64 vm.

**Comment by kmx** [comment]
The reason might be extra compiler option `-D__USE_MINGW_ANSI_STDIO` used with `longdouble` builds.

It is very likely that with `__USE_MINGW_ANSI_STDIO` inf and nan values are handled differently.

**Comment by kmx** [comment]
> Interesting thanks. I'll boot up my win64 vm.

Cool, in that case I suggest you to download http://strawberryperl.com/download/5.22.0.1/strawberry-perl-ld-5.22.0.1-64bit-portable.zip (there is all you need: gcc, gdb, dmake). You can demonstrate the failure by simply `cpanm Cpanel::JSON::XS -v`.

**Comment by rurban** [comment]
can fully reproduce it. thanks for the __USE_MINGW_ANSI_STDIO hint.

**Comment by rurban** [comment]
Fixed with 3.0211
